nautobot_ssot/metrics.py

```diff
@@ -0,0 +1,156 @@
+                    + last_job_sync.source_load_time.microseconds,
+                )
+
+            if last_job_sync.target_load_time:
+                ssot_job_durations.add_metric(
+                    labels=["target_load_time", job.slug],
+                    value=(last_job_sync.target_load_time.seconds * 1000000)
+                    + last_job_sync.target_load_time.microseconds,
+                )
+
+            if last_job_sync.diff_time:
+                ssot_job_durations.add_metric(
+                    labels=["diff_time", job.slug],
+                    value=(last_job_sync.diff_time.seconds * 1000000) + last_job_sync.diff_time.microseconds,
+                )
+
+            if last_job_sync.sync_time:
+                ssot_job_durations.add_metric(
+                    labels=["sync_time", job.slug],
+                    value=(last_job_sync.sync_time.seconds * 1000000) + last_job_sync.sync_time.microseconds,
+                )
+
+            if last_job_sync.duration:
+                ssot_job_durations.add_metric(
+                    labels=["sync_duration", job.slug],
+                    value=(last_job_sync.duration.seconds * 1000000) + last_job_sync.duration.microseconds,
+                )
+
+    yield ssot_job_durations
+
+
+def metric_syncs():
+    """Calculate total number of Syncs.
+
+    Yields:
+        GaugeMetricFamily: Prometheus Metrics
+    """
+    sync_gauge = GaugeMetricFamily("nautobot_ssot_sync_totals", "Nautobot SSoT Sync Totals", labels=["sync_type"])
+
+    sync_gauge.add_metric(labels=["total_syncs"], value=Sync.objects.all().count())
+
+    for status_type in ("completed", "failed", "errored", "pending", "running"):
+        sync_gauge.add_metric(
+            labels=[f"{status_type}_syncs"], value=Sync.objects.filter(job_result__status=status_type).count()
+        )
+
+    yield sync_gauge
+
+
+def metric_sync_operations():
+    """Extracts the diff summary operations from each Job's last Sync.
+
+    Yields:
+        GuageMetricFamily: Prometheus Metrics
+    """
+    sync_ops = GaugeMetricFamily(
+        "nautobot_ssot_sync_operations", "Nautobot SSoT operations by Job", labels=["job", "operation"]
+    )
+
+    for job in Job.objects.filter(slug__icontains="ssot"):
+        last_job_sync = Sync.objects.filter(job_result__job_model_id=job.id).last()
+        if last_job_sync and last_job_sync.summary:
+            for operation, value in last_job_sync.summary.items():
+                sync_ops.add_metric(
+                    labels=[job.slug, operation],
+                    value=value,
+                )
+
+    yield sync_ops
+
+
+def metric_memory_usage():
+    """Extracts memory usage for latest SSoT Sync where memory profiling was used.
+
+    Yields:
+        GaugeMetricFamily: Prometheus Metrics
+    """
+    memory_gauge = GaugeMetricFamily(
+        "nautobot_ssot_sync_memory_usage", "Nautobot SSoT Sync Memory Usage", labels=["phase"]
```

Comment:
Should this have a job label?

Reply:
```suggestion
        "nautobot_ssot_sync_memory_usage_bytes", "Nautobot SSoT Sync Memory Usage", labels=["phase"]
```

Reply:
We can do that, sure. I wasn't sure how applicable that'd be. My understanding is that you'd most likely use memory profiling for troubleshooting purposes so it wouldn't be enabled on Jobs all the time and you wouldn't necessarily care about it for all Jobs.

Reply:
I think this is still interesting to know. If I am building a dashboard about it, the data is a lot more comparable between different jobs if I have this label. Otherwise I can only graph the different jobs in a single graph.

Reply:
Refactored to include Job as a label.